kbp/config/acl_checker.go

```diff
@@ -0,0 +1,198 @@
+	read bool
+	list bool
+}
+
+// username -> permissionsV1
+type accessControlV1 struct {
+	whitelistAdditional map[string]permissionsV1
+	anonymous           permissionsV1
+}
+
+func parsePermissionsV1(permsStr string) (permissionsV1, error) {
+	permsStr = strings.TrimSpace(permsStr)
+	var perms permissionsV1
+	if len(permsStr) == 0 {
+		return perms, nil
+	}
+	for _, p := range strings.Split(permsStr, ",") {
+		switch p {
+		case "read":
+			perms.read = true
+		case "list":
+			perms.list = true
+		default:
+			return permissionsV1{}, ErrInvalidPermissions{
+				permissions: permsStr}
+		}
+	}
+	return perms, nil
+}
+
+func (ac *AccessControlV1) makeAccessControlV1Internal(users map[string]UserV1) (
+	aci *accessControlV1, err error) {
+	if ac == nil {
+		return nil, errors.New("nil AccessControlV1")
+	}
+	aci = &accessControlV1{}
+	aci.anonymous, err = parsePermissionsV1(ac.AnonymousPermissions)
+	if err != nil {
+		return nil, err
+	}
+	for username, permissions := range ac.WhitelistAdditionalPermissions {
+		if _, ok := users[username]; !ok {
+			return nil, ErrUndefinedUsername{username: username}
+		}
+		if aci.whitelistAdditional == nil {
+			aci.whitelistAdditional = make(map[string]permissionsV1)
+		}
+		aci.whitelistAdditional[username], err = parsePermissionsV1(permissions)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return aci, nil
+}
+
+type aclCheckerV1 struct {
+	children map[string]*aclCheckerV1
+	aci      *accessControlV1
+}
+
+// cleanPath cleans thePath in following order:
+//   1) Call path.Clean;
+//   2) Remove leading and trailing "/";
+//   3) Split the rest.
+// If thePath is root, nil elems is returned. Otherwise, elems is guaranteed to
+// be non-empty.
+//
+// If split2 is true, at most elems has at most 2 elements and the second one
+// would include everything other than the first element in the path;
+// otherwise, the entire thePath is split by "/" till the end of the world.
+func cleanPath(thePath string, split2 bool) (elems []string) {
+	thePath = path.Clean(thePath)
+	if thePath == "/" || thePath == "." {
+		// Examples of thePath (before clean) that lead us here:
+		//   ""
+		//   "/"
+		//   "."
+		//   "/.."
+		//   "/."
+		return nil
+	}
+	if strings.HasPrefix(thePath, "/") {
+		thePath = thePath[1:]
+	}
+	if strings.HasSuffix(thePath, "/") {
+		thePath = thePath[:len(thePath)-1]
+	}
+	// Now thePath can only be form of "a/b/c", i.e., no leading or trailing
+	// "/". Examples:
+	//   "a"
+	//   "a/b"
+	//   "a/b/c"
+	if split2 {
+		return strings.SplitN(thePath, "/", 2)
+	}
+	return strings.Split(thePath, "/")
+}
+
+// getAccessControl gets the corresponding accessControlV1 for thePath.
+// If a specifically defined exists, it's returned. Otherwise the default one
+// is returned.
+func (m *aclCheckerV1) getAccessControl(thePath string) (aci *accessControlV1) {
+	elems := cleanPath(thePath, true)
+	if elems == nil || m.children == nil {
+		return m.aci
+	}
+	// elems can't be empty.
```

Comment:
Technically it still could be empty if `cleanPath` has a bug:  https://play.golang.org/p/SsOPVoFvy6

Maybe better to replace `elems == nil` with `len(elems) == 0` above?

Reply:
Oh I meant cleanPath shouldn't generate empty `elems` because `strings.Split` never returns an empty slice unless the separator is empty. Will clarify the comment.

Reply:
Thanks.  Though I still think the `len()` check would be safer above.

Reply:
Sure! will do